kubetest/kind.go

```diff
@@ -0,0 +1,177 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"errors"
+	"flag"
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+)
+
+// This deployer requires --root=/go/src when pulling repositories.
+const (
+	kindBinaryPath = "/go/bin"
+	kindSrcPath    = "/go/src/sigs.k8s.io/kind"
+	k8sSrcPath     = "/go/src/k8s.io/kubernetes"
+	kindNodeImage  = "kindest/node:latest"
+)
+
+var (
+	kindConfigPath = flag.String("kind-config-path", "",
+		"(kind only) Path to the kind configuration file.")
+)
+
+// Definition for the kind deployer.
+type kind struct {
+	configPath string
+}
+
+// newKind creates a new kind deployer.
+func newKind() (deployer, error) {
+	k, err := initializeKind()
+	if err != nil {
+		return nil, err
+	}
+	return k, nil
+}
+
+// initializeKind initializers the kind deployer flags.
+func initializeKind() (*kind, error) {
+	// Ensure the kind binary path is added in $PATH.
+	// Could be redundant if kindBinaryPath is /go/bin and this is already in $PATH.
+	path := os.Getenv("PATH")
+	if !strings.Contains(path, kindBinaryPath) {
+		if err := os.Setenv("PATH", path+":"+kindBinaryPath); err != nil {
+			return nil, err
+		}
+	}
+	k := &kind{
+		configPath: *kindConfigPath,
+	}
+	return k, nil
+}
+
+func getKubeConfigPath() (string, error) {
+	o, err := control.Output(exec.Command("kind", "get", "kubeconfig-path"))
+	if err != nil {
+		return "", err
+	}
+	return strings.TrimSuffix(string(o), "\n"), nil
+}
+
+func setKubeConfigEnv() error {
+	path, err := getKubeConfigPath()
+	if err != nil {
+		return err
+	}
+	if err = os.Setenv("KUBECONFIG", path); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Builds kind an the node image and deploys a cluster based on the kind config.
+func (k *kind) Up() error {
+	// Build the kind binary.
```

Comment:
let's add a TODO to fetch a stable release binary and stash it in a kubetest specific dir.
we can add some mechanism to override to a different local kind binary for development, but default to fetching the stable release for the host if possible 🤔